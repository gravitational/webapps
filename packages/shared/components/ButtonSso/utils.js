/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import * as Icons from 'design/Icon';
import { AuthProviderTypeEnum } from '../../services/enums';

export const TypeEnum = {
  MICROSOFT: 'microsoft',
  GITHUB: 'github',
  BITBUCKET: 'bitbucket',
  GOOGLE: 'google',
};

export function pickSsoIcon(type) {
  switch (type) {
    case TypeEnum.MICROSOFT:
      return { color: '#2672ec', Icon: Icons.Windows, type };
    case TypeEnum.GITHUB:
      return { color: '#444444', Icon: Icons.Github, type };
    case TypeEnum.BITBUCKET:
      return { color: '#205081', Icon: Icons.BitBucket, type };
    case TypeEnum.GOOGLE:
      return { color: '#dd4b39', Icon: Icons.Google, type };
    default:
      // provide default icon for unknown social providers
      return { color: '#f7931e', Icon: Icons.OpenID };
  }
}

export function guessProviderType(name, ssoType) {
  name = name.toLowerCase();

  if (name.indexOf(TypeEnum.MICROSOFT) !== -1) {
    return TypeEnum.MICROSOFT;
  }

  if (name.indexOf(TypeEnum.BITBUCKET) !== -1) {
    return TypeEnum.BITBUCKET;
  }

  if (name.indexOf(TypeEnum.GOOGLE) !== -1) {
    return TypeEnum.GOOGLE;
  }

  if (
    name.indexOf(TypeEnum.GITHUB) !== -1 ||
    ssoType === AuthProviderTypeEnum.GITHUB
  ) {
    return TypeEnum.GITHUB;
  }

  if (ssoType === AuthProviderTypeEnum.OIDC) {
    return 'openid';
  }

  return '--unknown';
}
