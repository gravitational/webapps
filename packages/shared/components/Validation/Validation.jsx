/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import Logger from '../../libs/logger';
import { isObject } from 'lodash';

const logger = Logger.create('validation');

// Validator handles input validation
export default class Validator {
  valid = true;

  constructor() {
    // store subscribers
    this._subs = [];
  }

  // adds a callback to the list of subscribers
  subscribe(cb) {
    this._subs.push(cb);
  }

  // removes a callback from the list of subscribers
  unsubscribe(cb) {
    const index = this._subs.indexOf(cb);
    if (index > -1) {
      this._subs.splice(index, 1);
    }
  }

  addResult(result) {
    // result can be a boolean value or an object
    let isValid = false;
    if (isObject(result)) {
      isValid = result.valid;
    } else {
      logger.error(`rule should return a valid object`);
    }

    this.valid = this.valid && Boolean(isValid);
  }

  reset() {
    this.valid = true;
    this.validating = false;
  }

  validate() {
    this.reset();
    this.validating = true;
    this._subs.forEach(cb => {
      try {
        cb();
      } catch (err) {
        logger.error(err);
      }
    });

    return this.valid;
  }
}

const ValidationContext = React.createContext({});

export function Validation(props) {
  const [validator] = React.useState(() => new Validator());
  // handle render functions
  const children =
    typeof props.children === 'function'
      ? props.children({ validator })
      : props.children;

  return (
    <ValidationContext.Provider value={validator}>
      {children}
    </ValidationContext.Provider>
  );
}

export function useValidation() {
  const value = React.useContext(ValidationContext);
  if (!(value instanceof Validator)) {
    logger.warn('Missing Validation Context declaration');
  }

  return value;
}
