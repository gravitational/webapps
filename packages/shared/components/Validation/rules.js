/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * requiredField checks for empty strings and arrays.
 *
 * @param message The custom error message to display to users.
 * @param value The value user entered.
 */
const requiredField = message => value => () => {
  const valid = !(!value || value.length === 0);
  return {
    valid,
    message: !valid ? message : '',
  };
};

const requiredToken = value => () => {
  if (!value || value.length === 0) {
    return {
      valid: false,
      message: 'Token is required',
    };
  }

  return {
    valid: true,
  };
};

const requiredPassword = value => () => {
  if (!value || value.length < 6) {
    return {
      valid: false,
      message: 'Enter at least 6 characters',
    };
  }

  return {
    valid: true,
  };
};

const requiredConfirmedPassword = password => confirmedPassword => () => {
  if (!confirmedPassword) {
    return {
      valid: false,
      message: 'Please confirm your password',
    };
  }

  if (confirmedPassword !== password) {
    return {
      valid: false,
      message: 'Password does not match',
    };
  }

  return {
    valid: true,
  };
};

export {
  requiredToken,
  requiredPassword,
  requiredConfirmedPassword,
  requiredField,
};
