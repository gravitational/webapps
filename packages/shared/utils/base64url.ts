/**
 * Copyright 2021 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export type Base64urlString = string;

// Sourced from https://github.com/github/webauthn-json
export function base64urlToBuffer(
  baseurl64String: Base64urlString
): ArrayBuffer {
  // Base64url to Base64
  const padding = '=='.slice(0, (4 - (baseurl64String.length % 4)) % 4);
  const base64String =
    baseurl64String.replace(/-/g, '+').replace(/_/g, '/') + padding;

  // Base64 to binary string
  const str = atob(base64String);

  // Binary string to buffer
  const buffer = new ArrayBuffer(str.length);
  const byteView = new Uint8Array(buffer);
  for (let i = 0; i < str.length; i++) {
    byteView[i] = str.charCodeAt(i);
  }
  return buffer;
}

export function bufferToBase64url(buffer: ArrayBuffer): Base64urlString {
  // Buffer to binary string
  const str = String.fromCharCode.apply(null, new Uint8Array(buffer));

  // Binary string to base64
  const base64String = btoa(str);

  // Base64 to base64url
  // We assume that the base64url string is well-formed.
  const base64urlString = base64String
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
  return base64urlString;
}
