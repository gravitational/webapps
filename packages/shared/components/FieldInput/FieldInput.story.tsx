/*
Copyright 2022 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';

import Validation from '../../components/Validation';

import FieldInput from './FieldInput';

export default {
  title: 'Shared',
};

export const Fields = () => (
  <Validation>
    {() => (
      <>
        <FieldInput
          mb="6"
          label="Label"
          labelTip="Optional tabel tip"
          name="optional name"
          onChange={() => {}}
          value={'value'}
        />
        <FieldInput
          mb="6"
          label="Label with placeholder"
          name="optional name"
          onChange={() => {}}
          placeholder="placeholder"
          value={''}
        />
        <FieldInput
          mb="6"
          placeholder="without label"
          validator={() => false}
          onChange={() => {}}
        />
      </>
    )}
  </Validation>
);

Fields.storyName = 'FieldInput';
