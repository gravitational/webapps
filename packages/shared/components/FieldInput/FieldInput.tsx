/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { Box, Input, LabelInput } from 'design';
import { useRule } from 'shared/components/Validation';

export default function FieldInput({
  label,
  value,
  onChange,
  onKeyPress,
  placeholder,
  rule = defaultRule,
  type = 'text',
  autoFocus = false,
  autoComplete = 'off',
  readonly = false,
  ...styles
}: Props) {
  const { valid, message } = useRule(rule(value));
  const hasError = !valid;
  const labelText = hasError ? message : label;
  return (
    <Box mb="4" {...styles}>
      {label && <LabelInput hasError={hasError}>{labelText}</LabelInput>}
      <Input
        type={type}
        autoFocus={autoFocus}
        hasError={hasError}
        placeholder={placeholder}
        value={value}
        autoComplete={autoComplete}
        onChange={onChange}
        onKeyPress={onKeyPress}
        readOnly={readonly}
      />
    </Box>
  );
}

const defaultRule = () => () => ({ valid: true });

type Props = {
  value?: string;
  label?: string;
  placeholder?: string;
  autoFocus?: boolean;
  autoComplete?: 'off' | 'on';
  type?: 'email' | 'text' | 'password' | 'number' | 'date' | 'week';
  rule?: Function;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onKeyPress?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
  readonly?: boolean;
  // TS: temporary handles ...styles
  [key: string]: any;
};
