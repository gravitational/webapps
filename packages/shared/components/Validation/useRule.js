/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import Logger from '../../libs/logger';

import { useValidation } from './Validation';

const logger = Logger.create('validation');

/**
 * useRule subscribes to validation requests upon which executes validate() callback
 */
export default function useRule(cb) {
  if (typeof cb !== 'function') {
    logger.warn(`useRule(fn), fn() must be a function`);
    return;
  }

  const [, rerender] = React.useState();
  const validator = useValidation();

  // register to validation context to be called on cb()
  React.useEffect(() => {
    function onValidate() {
      if (validator.validating) {
        const result = cb();
        validator.addResult(result);
        rerender({});
      }
    }

    // subscribe to store changes
    validator.subscribe(onValidate);

    // unsubscribe on unmount
    function cleanup() {
      validator.unsubscribe(onValidate);
    }

    return cleanup;
  }, [cb]);

  // if validation has been requested, cb right away.
  if (validator.validating) {
    return cb();
  }

  return { valid: true };
}
