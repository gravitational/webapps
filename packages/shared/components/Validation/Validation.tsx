/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import Logger from 'shared/libs/logger';
import { RuleReturnType } from 'shared/components/Validation/rules';
import { isObject } from 'lodash';

const logger = Logger.create('validation');

// Validator handles input validation
export default class Validator {
  valid = true;
  validating = false;
  private _subs: Function[];

  constructor() {
    // store subscribers
    this._subs = [];
  }

  // adds a callback to the list of subscribers
  subscribe(cb: Function) {
    this._subs.push(cb);
  }

  // removes a callback from the list of subscribers
  unsubscribe(cb: Function) {
    const index = this._subs.indexOf(cb);
    if (index > -1) {
      this._subs.splice(index, 1);
    }
  }

  addResult(result: RuleReturnType) {
    // result can be a boolean value or an object
    let isValid = false;

    // TODO: remove check when all files using Validation is converted to TS
    if (isObject(result)) {
      isValid = result.valid;
    } else {
      logger.error(`rule should return a valid object`);
    }

    this.valid = this.valid && Boolean(isValid);
  }

  reset() {
    this.valid = true;
    this.validating = false;
  }

  validate() {
    this.reset();
    this.validating = true;
    this._subs.forEach(cb => {
      try {
        cb();
      } catch (err) {
        logger.error(err);
      }
    });

    return this.valid;
  }
}

const ValidationContext = React.createContext<Partial<Validator>>({});

export function Validation(props: { children: React.ReactNode }) {
  const [validator] = React.useState(() => new Validator());
  // handle render functions
  const children =
    typeof props.children === 'function'
      ? props.children({ validator })
      : props.children;

  return (
    <ValidationContext.Provider value={validator}>
      {children}
    </ValidationContext.Provider>
  );
}

export function useValidation() {
  const value = React.useContext(ValidationContext);
  if (!(value instanceof Validator)) {
    logger.warn('Missing Validation Context declaration');
  }

  return value;
}
