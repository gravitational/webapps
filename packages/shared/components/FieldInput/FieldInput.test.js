/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import FieldInput from './FieldInput';
import { render, fireEvent } from 'design/utils/testing';
import { useRule } from '../Validation';
import theme from '../../../design/src/theme';

jest.mock('./../Validation/useRule');

test('valid values, autofocus, onChange, onKeyPress', () => {
  const rule = jest.fn();
  const onChange = jest.fn();
  const onKeyPress = jest.fn();

  // mock positive validation
  useRule.mockImplementation(() => ({ valid: true, message: '' }));

  const { getByText, getByPlaceholderText } = render(
    <FieldInput
      placeholder="placeholderText"
      autoFocus={true}
      label="labelText"
      rule={rule}
      onChange={onChange}
      onKeyPress={onKeyPress}
    />
  );

  // test label is displayed
  expect(getByText('labelText')).toBeInTheDocument();

  // test autofocus prop is respected
  const inputEl = getByPlaceholderText('placeholderText');
  expect(document.activeElement).toEqual(inputEl);

  // test onChange prop is respected
  fireEvent.change(inputEl, { target: { value: 'test' } });
  expect(onChange).toHaveBeenCalledTimes(1);

  // test onKeyPress prop is respected
  fireEvent.keyPress(inputEl, { key: 'Enter', keyCode: 13 });
  expect(onKeyPress).toHaveBeenCalledTimes(1);
});

test('input validation error state', () => {
  const rule = jest.fn();
  const errorColor = theme.colors.error.main;

  // mock negative validation
  useRule.mockImplementation(() => ({ valid: false, message: 'errorMsg' }));

  const { getByText, getByPlaceholderText } = render(
    <FieldInput placeholder="placeholderText" label="labelText" rule={rule} />
  );

  // test !valid values renders with error message
  const labelEl = getByText('errorMsg');
  expect(labelEl).toHaveStyle({ color: errorColor });

  // test !valid values renders error colors
  const inputEl = getByPlaceholderText('placeholderText');
  expect(inputEl).toHaveStyle({
    'border-color': errorColor,
  });
});
