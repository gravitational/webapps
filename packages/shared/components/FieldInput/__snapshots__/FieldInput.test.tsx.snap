// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests 1`] = `
.c0 {
  box-sizing: border-box;
  margin-bottom: 40px;
}

.c6 {
  display: inline-block;
  transition: color 0.3s;
  color: #FFFFFF;
  font-size: 18px;
}

.c3 {
  appearance: none;
  border: none;
  border-radius: 4px;
  box-shadow: inset 0 2px 4px rgba(0,0,0,0.24);
  box-sizing: border-box;
  display: block;
  height: 40px;
  font-size: 16px;
  padding: 0 16px;
  outline: none;
  width: 100%;
  color: #324148;
  background-color: #FFFFFF;
  margin-top: 4px;
  border: 2px solid #ff1744;
  padding: 10px 14px;
}

.c3::-ms-clear {
  display: none;
}

.c3::placeholder {
  opacity: 0.4;
}

.c3:read-only {
  cursor: not-allowed;
}

.c1 {
  color: #ff1744;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 0px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0px;
}

.c5 {
  vertical-align: middle;
  display: inline-block;
  height: 18px;
}

.c5:hover {
  cursor: pointer;
}

.c4 {
  margin-right: 4px;
  vertical-align: middle;
}

<div>
  <div
    class="c0"
    name="optional name"
  >
    <label
      class="c1"
      font-size="0"
    >
      errorMsg
      <span
        class="c2"
        style="font-weight: normal;"
      >
         - Optional tabel tip
      </span>
      <input
        autocomplete="off"
        class="c3"
        color="text.onLight"
        inputmode="text"
        type="text"
        value="value"
      />
    </label>
  </div>
  <div
    class="c0"
    name="optional name"
  >
    <label
      class="c1"
      font-size="0"
    >
      errorMsg
      <input
        autocomplete="off"
        class="c3"
        color="text.onLight"
        inputmode="text"
        placeholder="placeholder"
        type="text"
        value=""
      />
    </label>
  </div>
  <div
    class="c0"
    name="optional name"
  >
    <label
      class="c1"
      font-size="0"
    >
      <span
        class="c4"
      >
        errorMsg
      </span>
      <span
        class="c5"
      >
        <span
          class="c6 icon icon-info_outline "
          color="light"
          font-size="4"
        />
      </span>
      <input
        autocomplete="off"
        class="c3"
        color="text.onLight"
        inputmode="text"
        placeholder="placeholder"
        type="text"
        value=""
      />
    </label>
  </div>
  <div
    class="c0"
    name="optional name"
  >
    <label
      class="c1"
      font-size="0"
    >
      <span
        class="c4"
      >
        errorMsg
        <span
          class="c2"
          style="font-weight: normal;"
        >
           - the label tip
        </span>
      </span>
      <span
        class="c5"
      >
        <span
          class="c6 icon icon-info_outline "
          color="light"
          font-size="4"
        />
      </span>
      <input
        autocomplete="off"
        class="c3"
        color="text.onLight"
        inputmode="text"
        placeholder="placeholder"
        type="text"
        value=""
      />
    </label>
  </div>
  <div
    class="c0"
  >
    <input
      autocomplete="off"
      class="c3"
      color="text.onLight"
      inputmode="text"
      placeholder="without label"
      type="text"
      value=""
    />
  </div>
</div>
`;
